working on logic.

yanhao-li opened the pull request:
we can now support the following logic flows

if () {...}
if () {...} else {...}
(expr) ? true stmt; : false stmt;

To facilitate this I needed to add a No Op where if there is no else
branch like above we create an AST for if(predicate, true block, No_op).

This doesnt break anything and there are no shift/reduce errors.

We just need to check if s2, in if(predicate, s1, s2) is a No_op when
doing code gen.

Further, look at test/parser/logic for examples. It passes the -a
and -s checks

Matt20D committed:
working on logic.

we can now support the following logic flows

if () {...}
if () {...} else {...}
(expr) ? true stmt; : false stmt;

To facilitate this I needed to add a No Op where if there is no else
branch like above we create an AST for if(predicate, true block, No_op).

This doesnt break anything and there are no shift/reduce errors.

We just need to check if s2, in if(predicate, s1, s2) is a No_op when
doing code gen.

Further, look at test/parser/logic for examples. It passes the -a
and -s checks

yanhao-li commented on lib/semant.ml:
```diff
@@ -212,6 +216,7 @@ let check (globals, functions) =
         else raise (
             Failure ("return gives " ^ string_of_typ t ^ " expected " ^
                      string_of_typ func.rtyp ^ " in " ^ string_of_expr e))
+      | No_op -> SNo_op (* for the case where we only want if (..) {...} with no else block *)
```
we should just do it the way microc did it to handle the if with no else block.

Matt20D replied:
microC does not handle that case
compactiC does. I just dont want it to be like we "copied it"

we could just add a check in the SIf section, where if expr2 is No_op then we add the block there.

I just dont want it to be close to what compactiC does

Matt20D committed:
implemented INFINITE_LOOP

quite easy, just need to map it to the while construct that
we already employed

Matt20D committed:
changed around the if statements a little

added in an extra level of indirection needed for CFG.

needed to create a logic_expr rule that doesn't look for the semi
this is needed in for loop. So i changed it and made sure that
our if logic still works. check the test/programs/logic.viz file

Matt20D committed:
very big commit

we now have functionality for our for loop constructs.

we wrote in LRM that we wanted:

for counter in starting_num ... <ending condition>ending_num step step_number

That is implemented. The code builds with no shift/reduce errors
and my test/programs/loops.viz file will pass all checks.

run the test/programs/loops.viz file with -a and -s flags to double
check my work.

I think that this is correct. Semantic file also compiles. I essentially
mapped our for loop readable syntax, to that for a normal For(...)
loop

Matt20D committed:
test file for all of our loop constructs

while
infinite_loop
for loops

They all pass the semantic check, just waiting for a code review.

run this file with -a and -s flags to see the output.

yanhao-li commented on lib/fmt/token_fmt.ml:
```diff
@@ -62,7 +62,8 @@ let string_of_token (t : token) : string  =
     | DIVEQ -> "DIVEQ"
-    | MODEQ -> "MODEQ"       
+    | MODEQ -> "MODEQ"
+    | QUESTION -> "TERNARY_IF_STATEMENT"       
```
I suggest we format the token exactly as it's class name("QUESTION"), so that we could automatically generate this part in the future. 

and I believe ternary if statement is the only place `QUESTION` token will be used, so everyone knows what question means.

but it's a minor update, depends on you ;)

Matt20D replied:
yea true, I just wanted to be clear lol

yanhao-li commented on lib/parser.mly:
```diff
@@ -110,10 +110,66 @@ stmt_list:
-  | LBRACE stmt_list RBRACE                 { Block $2 }
-  | IF LPAREN expr RPAREN stmt ELSE stmt    { If($3, $5, $7) }
-  | WHILE LPAREN expr RPAREN stmt           { While ($3, $5)  }
+  | logic_expr                              { Expr $1 }
```
this line looks weird to me, all the statements (except control blocks) should end with SEMI, or it is not a statement.

Matt20D replied:
thats not true though

if (x < y) {

}

not if (x < y); {

}

do you see what I mean? only in logical expressions, i.e. booleans will this be the case where we use the Expr with no semi.

I noticed this in the ternary operator this is how java does it: 

booleanExpression ? expression1 : expression2

there shouldn't be a semi in any of those cases.

yanhao-li replied:
Yes, I understand why you added this `logic_expr`, and what you want to achieve totally makes sense. 
But it doesn't change the definition of a statement, if a clause is a statement, it supposes to end with semi, or it is a if/loop statement. 

We just need to figure out another way to achieve this.
I will say `expr QUESTION stmt COLON stmt` should be changed to `expr QUESTION expr COLON expr`, because the second and third clauses will return, so they are expression.

Matt20D replied:
sorry, I dont follow. maybe you could explain to me on discord

yanhao-li commented on lib/parser.mly:
```diff
@@ -110,10 +110,66 @@ stmt_list:
+  | WHILE expr stmt           { While ($2, $3)  }
+  | INFINITE_LOOP stmt        { While (BoolLit(true), $2)  }
+  /* for counter in starting_num ... <ending condition>ending_num step step_number */
+  | FOR ID_VAR IN LIT_INT DOT DOT DOT end_condition LIT_INT increment stmt
```
`DOT DOT DOT` should be parsed as `SPREAD` token. otherwise like `. .\n.` will be accepted.

Matt20D replied:
oh thats a good call, what is spread? is this a new token to be added?

yanhao-li replied:
yes, I believe so, `SPREAD` or `RANGE` sounds like a good token name for this.

Matt20D replied:
I will change it

yanhao-li commented on lib/ast.ml:
```diff
@@ -28,8 +28,10 @@ type stmt =
+  | For of expr * expr * expr * stmt
   | Return of expr
-
+  | No_op
```
`No_op` as a statement does not make much sense to me. I will prefer to avoid those ambiguous definitions.

Matt20D replied:
thats not ambiguous, in fact CompactiC implements that as noexpr.

it is a placeholder, for something that doesnt exist. I have added an if statement in the semantic code that handles this. In fact exactly how microC (Stephen Edward's) handles it in the CFG, but I do it at the semantics time.

It will not in fact cause codegen issues, because I add in an empty SBlock, and codegen uses the sast.

I know this is for machine code, but this reference will explain how its just a place holder: 
https://www.noop.com.au/noop-meaning/

There are no ambiguities in the grammar, or else there would be shift/reduce or reduce/reduce conflicts. All of that is being handled

yanhao-li replied:
`noexpr` is indeed needed to implement microc, but I believe they don't have `nostmt`, and it sounds not right to me. 

 I also don't think we need `noexpr` though. it's there in microc only because they support c's legacy grammar, i.e.
`for (;;) {}` and `return;`. 

But we don't have this kind of grammar, in viz, we are using `infinite_loop(){}` and `return none;`. no? 😄

I believe we can make it work in codegen, just seems redundant to me. Maybe we could ask others' opinions.

Matt20D committed:
handle the no_op

no_op is a place holder for an if without an else block.

So in the SAST if we see a SNo_op then we just replace it with an
empty SBlock([])

all of the loop constructs still pass the semantics check. And now,
there will be no issue with codegen
